crates/mo-file-parser/src/lib.rs

```diff
@@ -0,0 +1,122 @@
+            // differentiate between the two either.
+            // All features we care about are supported in minor revision 0,
+            // so we do not need to care about the minor revision.
+            Ok(())
+        }
+        _ => Err(std::io::Error::new(
+            std::io::ErrorKind::InvalidData,
+            "Major revision must be 0 or 1",
+        )),
+    }
+}
+
+fn parse_strings(
+    file_content: &[u8],
+    num_strings: usize,
+    table_offset: usize,
+    read_u32: fn(&[u8]) -> u32,
+) -> std::io::Result<Vec<&[u8]>> {
+    let file_too_short_error = Err(std::io::Error::new(
+        std::io::ErrorKind::InvalidData,
+        "MO file is too short.",
+    ));
+    if table_offset + num_strings * 2 * U32_SIZE > file_content.len() {
+        return file_too_short_error;
+    }
+    let mut strings = Vec::with_capacity(num_strings);
+    let mut offset = table_offset;
+    let mut get_next_u32 = || {
+        let val = read_u32(&file_content[offset..]);
+        offset += U32_SIZE;
+        val
+    };
+    for _ in 0..num_strings {
+        // not including NUL terminator
+        let string_length = get_next_u32() as usize;
+        let string_offset = get_next_u32() as usize;
+        if string_offset + string_length > file_content.len() {
+            return file_too_short_error;
+        }
+        // Contexts are stored by storing the concatenation of the context, a EOT byte, and the original string, instead of the original string.
+        // Contexts are not supported by this implementation.
+        // The format allows plural forms to appear behind singular forms, separated by a NUL byte,
+        // where `string_length` includes the length of both.
+        // This is not supported here.
+        // Do not include the NUL terminator in the slice.
+        strings.push(&file_content[string_offset..(string_offset + string_length)]);
+    }
+    Ok(strings)
+}
+
+/// Parse a MO file.
+/// Format reference used: <https://www.gnu.org/software/gettext/manual/html_node/MO-Files.html>
+pub fn parse_mo_file(file_content: &[u8]) -> std::io::Result<HashMap<&[u8], &[u8]>> {
+    if file_content.len() < 7 * U32_SIZE {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::InvalidData,
+            "File too short to contain header.",
+        ));
+    }
+    // The first 4 bytes are a magic number, from which the endianness can be determined.
+    let read_u32 = get_u32_reader_from_magic_number(&file_content[0..U32_SIZE])?;
+    let mut offset = U32_SIZE;
+    let mut get_next_u32 = || {
+        let val = read_u32(&file_content[offset..]);
+        offset += U32_SIZE;
+        val
+    };
+    let file_format_revision = get_next_u32();
+    check_if_revision_is_supported(file_format_revision)?;
+    let num_strings = get_next_u32() as usize;
+    let original_strings_offset = get_next_u32() as usize;
+    let translation_strings_offset = get_next_u32() as usize;
```

Comment:
I think we should replace ` as usize` with `usize::try_from().unwrap()`.
Ideally we can cause a compile time error if `u32::MAX > usize::MAX`,
and then minimize the syntax overhead, maybe by extracting a function